### ISSUE TITLE
Consider adding data as asset and making the repo lighter!

### ISSUE THREAD

**Issue by nicolas-heigvd** [opengeospatial/GeoPose#25]
Hello,

Can you check why the size of this repository is so large (>3.0 GB !)?

It is generally not a good idea to include large data _inside_ a git repo because it will stay in the git history for **ever**. For everyone. Also, you probably don't want to track changes to the data itself, especially if they are demo or sample data. Instead, it is much better to attach data as some asset with a release tag for example.

This large size currently makes the repo unusable for contributors willing to only work on the specs themselves.

Many thanks!

**Comment by 3DXScape** [comment]
Actually the data _is_ changing as the documents change and as errors or missing things are discovered in the data. You have a good point though, and we/I will come up with a way to make this lighter.

**Comment by cperey** [comment]
@3DXScape talked about this issue previously and steve has taken an action item to develop an upload feature to GeoPose.io (on Amazon), change the links in the repo to the data elsewhere. 

Experiments with the data sets are really helpful in revealing ways to use GeoPose.

If large data set has been in the GitHub repo, will that stay around? 

Nicolas/Steve: normally, yes you will need to reset it.

**Comment by cperey** [comment]
Nicolas: there are some tools but they are not straight-forward to use

**Comment by 3DXScape** [comment]
Linus T wants only smart people to use git.

**Comment by 3DXScape** [comment]
Per our last meeting discussion, I have implemented out a scheme for offloading binary large objects (blobs) to Amazon S3. I created a versioned public bucket named **geopose1data** with the same folder structure as the repo. To implement, a person moves a blob from a repo folder repo to the corresponding folder in the bucket. The person puts the S3 link to the blob into a new or existing **assets.adoc** file in the original repo folder. These links take you directly to the blob (and back via the browser back button/arrow). [Here](https://github.com/opengeospatial/GeoPose/tree/main/Data/Examples/ROS-GeoPose/Experiments/FFVideo) is a link that I have provided as a template. Unless I hear protests from the SWG, I will move most of the blobs (excluding some static individual image files) to S3 over the next few days. This will minimize growth of the repo and later, after some history hacking, it should be much smaller.

As for backups, I will make a zip every week and post it somewhere TBD.

**Comment by 3DXScape** [comment]
First feedback I have gotten leads me to think that we need a web UI for adding data files. As it is, even though the S3 bucket is public, you have to have an Amazon account to access it.  There's no direct cost but still it is a level of complexity that we can/should avoid. We need a façade that wraps the access methods.  I will add that shortly.

**Comment by nicolas-heigvd** [comment]
A quick question then: do you know if other people @OGC also routinely share large documents to work with during the development of a standard? Are there some common "good" practices regarding that?

In addition, there is this little something that tickles my ears; Amazon is a private company having for sure they own conditions, and that's fine if people agree with that for sharing some documents (with in the background questions about licenses, or ownership rights, etc...), but in fact, there is this tiny little "idealistic" gap between that and the 'O' -- standing for Open -- in OGC.

**Comment by jyutzler** [comment]
> A quick question then: do you know if other people @ OGC also routinely share large documents to work with during the development of a standard? Are there some common "good" practices regarding that?
> 
> In addition, there is this little something that tickles my ears; Amazon is a private company having for sure they own conditions, and that's fine if people agree with that for sharing some documents (with in the background questions about licenses, or ownership rights, etc...), but in fact, there is this tiny little "idealistic" gap between that and the 'O' -- standing for Open -- in OGC.

Short answer: yes. There are several ways this can be done.

* For in-progress documents, there is a GitHub a process that processes the AsciiDoc and posts it to a web-accessible location.
* For binary files (< 200 MB), there is the OGC Portal. Directories are organized by working group. Files can be made public or private depending on working group preference.
* For large binary files (> 200 MB), there is an OGC FTP server.
* I'm probably forgetting something.

**Comment by 3DXScape** [comment]
I have removed 1,400+ binary files and put them in the geopose1data bucket in Amazon S3. For access, I have automated the creation of assets.adoc files for every folder that has S3 content. I included an assets_info.adoc file as well but once I discovered that GitHub does not support adoc includes, it became less useful. The idea was that the automated process could keep the lists of binary files in sync automatically but the description of the category could be maintained separately in GitHub and be included automatically in the assets.adoc files. Apparently this limitation of GitHub has been the subject of a 3+ year long discussion with no indication if/when it will be added.

**Comment by 3DXScape** [comment]
Still to do:

* Add web UI to upload to S3 without having to deal with Amazon [ETA 15 July]
* Hack the history of the binary files out of the repo [ETA ???]

**Comment by 3DXScape** [comment]
OGC set up a new FTP site for us. I have set up the GitHub Data folder tree structure on the new FTP site. I have done an initial sync with S3. I modified the automation to generate links to the OGC location and updated the associated assets.adoc files. I will not have a chance to verify that all the new links and details are correct before our Friday meeting but it looks generally correct. Remaining steps are

1. to point the automation internally to the OSC mirror of the S3 content and
1. to provide a UI that make uploading new large binary files as simple as possible. 

Finishing these steps  is a couple of weeks away, given my current schedule. One thing to note if you enter a link manually (not sure when this would happen, but worth knowing if it comes up), you will find that the links are case-sensitive. This is a characteristic of the FTP site and should normally not matter since the links are automatically generated.
